### PR TITLE
Fixes #440 : Add Status to the available subscription field filters

### LINKED ISSUE #440: Add Status to the available subscription field filters
#### Summary

Currently, the field filters do not allow for a subscription such as "when a ticket's status is x".

#### Additional Info

The option needs to be added to the dropdown on the frontend for it to be supported. The change will likely need to go in `webapp/src/utils/jira_issue_metadata.tsx`

#### Screenshots

image

### PR DESCRIPTION
**Pull request by M-ZubairAhmed**
#### Summary
:construction:  WIP :construction: 
This PR adds the ability to filter out jira issue by status fields.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-plugin-jira/issues/440

### CODE REVIEW

**Comment by jfrerich** [comment]
Thanks for another PR, @M-ZubairAhmed! 

@mickmister, I noticed in the GH Issue for this PR you mentioned collecting the status values from the get-create-issue-metadata.  Looking at the available data, I don't think the available status values are included. Unless they are available, it seems the route that @M-ZubairAhmed created (server-side) will be needed.  What do you think?

**Comment by mickmister** [comment]
>  I don't think the available status values are included. Unless they are available, it seems the route that @M-ZubairAhmed created (server-side) will be needed.

@jfrerich This is correct. I discussed with @M-ZubairAhmed offline about this, and the [`StatusCategories`](https://moduscreate.com/blog/jira-status) that are being fetched in this PR are not sufficient, and we will need to fetch the `Statuses` instead. The Jira client we are using does not have a method for fetching `Statuses`, so we will need to use `client.Jira.Do` to make the request.

**Comment by M-ZubairAhmed** [comment]
Thank you all for clarification, i will get on this PR soon after hackathon.

**Comment by mickmister** [comment]
If we update the `go-jira` dependency, a `GetAllStatuses` method is available. Let's go with that https://github.com/andygrunwald/go-jira/releases/tag/v1.11.0

**Comment by M-ZubairAhmed** [comment]
@mickmister ok got it

**Comment by mickmister** [comment]
Let's upgrade to the latest version while we're updating https://github.com/andygrunwald/go-jira/releases

Thank you @M-ZubairAhmed, this is a great feature!

**Comment by M-ZubairAhmed** [comment]
@mickmister bumped the go-jira version :arrow_up:

File: server/client.go [hunk 1]
```diff
@@ -238,6 +240,15 @@ func (client JiraClient) UpdateComment(issueKey string, comment *jira.Comment) (
 	return updated, err
 }
 
+// GetAllStatuses returns a list of all statuses associated with JIRA workflows.
+func (client JiraClient) GetAllStatuses() ([]jira.Status, error) {
```
[the remark this answers is not in the document]
Comment (M-ZubairAhmed) [reply server/client.go]: Sure, done

File: server/http.go [hunk 2]
```diff
@@ -25,6 +25,7 @@ const (
 	routeAPIGetJiraProjectMetadata = "/api/v2/get-jira-project-metadata"
 	routeAPIGetSearchIssues        = "/api/v2/get-search-issues"
 	routeAPIAttachCommentToIssue   = "/api/v2/attach-comment-to-issue"
+	routeAPIGetAllStatuses         = "/api/v2/get-all-statuses"
```
[the remark this answers is not in the document]
Comment (M-ZubairAhmed) [reply server/http.go]: resolved

File: webapp/src/actions/index.js [hunk 3]
```diff
@@ -46,6 +46,28 @@ export const closeAttachCommentToIssueModal = () => {
     };
 };
 
+/**
+ * Returns list of statuses the jira project uses, list is not stored in store but returned from function
+ * @function fetchJiraProjectStatuses
+ * @param none
+ * @returns {promise} Promise object represents list data or error
```
[the remark this answers is not in the document]
Comment (M-ZubairAhmed) [reply webapp/src/actions/index.js]: resolved

File: server/issue.go [hunk 4]
```diff
@@ -572,6 +572,53 @@ func httpAPIAttachCommentToIssue(ji Instance, w http.ResponseWriter, r *http.Req
...
+	}
+
+	// Marshalling Statuses received from API to JSON
+	statusesBytes, err := json.Marshal(statuses)
+	if err != nil {
+		return http.StatusInternalServerError, errors.WithMessage(err, "failed to marshal response of Jira statuses")
+	}
+
+	// Prepare Header to send to Plugin WebClient
+	w.Header().Set("Content-Type", "application/json")
```
[the remark this answers is not in the document]
Comment (M-ZubairAhmed) [reply server/issue.go]: resolved

**Comment by jfrerich** [comment]
@mickmister, @M-ZubairAhmed, can we convert this from `draft` to a formal PR review?

**Comment by M-ZubairAhmed** [comment]
hi @jfrerich i still have to do one thing which @mickmister proposed

File: server/client.go [hunk 5]
```diff
@@ -238,6 +240,15 @@ func (client JiraClient) UpdateComment(issueKey string, comment *jira.Comment) (
 	return updated, err
 }
 
+// GetStatuses returns a list of all statuses associated with JIRA workflows.
+func (client JiraClient) GetStatuses() ([]jira.Status, error) {
```
Comment (levb) [inline server/client.go:244]: Can you please clarify: are these all statuses for the "instance"? "project"? it's not obvious how this request is contextualized. Maybe an API doc link?

**Comment by jasonblais** [comment]
Hey @M-ZubairAhmed  - I know it has been a while. Are you interested continuing to work on this pull request? Looks like it's close. cc @jfrerich

**Comment by M-ZubairAhmed** [comment]
Will continue after the Hackathon

**Comment by smorri1967** [comment]
Any update on when this might be available?

**Comment by M-ZubairAhmed** [comment]
Will continue work when i have the bandwidth :)

**Pull request closed without merging**